### ISSUE TITLE
Linker error (unplaced orphan section) with newer toolchains?

### ISSUE THREAD

**Issue by TobiX** [frno7/iopmod#3]
Hi,

I'm getting the following linker error when compiling with a current Gentoo toolchain (built today):
```
  LD      module/printk.iop
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `module/printk.o'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `module/printk.mod.o'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(printk.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(strlen.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(sif.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(vsnprintf.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(__udivdi3.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(memcpy.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(__umoddi3.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(__udivmoddi4.o)'
make: *** [Makefile:150: module/printk.iop] Error 1
  LD      module/irqrelay.iop
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `module/irqrelay.o'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `module/irqrelay.mod.o'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(interrupt.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(ioperr.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(sif.o)'
mipsr5900el-unknown-linux-gnu-ld: error: unplaced orphan section `.note.GNU-stack' from `builtin/builtin.a(memcpy.o)'
make: *** [Makefile:150: module/irqrelay.iop] Error 1
```

Versions:
- mipsr5900el-unknown-linux-gnu-ld GNU ld (Gentoo 2.34 p4) 2.34.0
- mipsr5900el-unknown-linux-gnu-gcc (Gentoo 10.1.0 p1) 10.1.0

Switching to `--orphan-handling=discard` seems to get past this problem, but I'm not sure if the resulting binaries work (still working on getting my system to boot :D)

**Comment by frno7** [comment]
Thanks for your report, @TobiX! I’m currently using `GNU ld (Gentoo 2.33.1 p2) 2.33.1` and `mipsr5900el-unknown-linux-gnu-gcc (Gentoo 9.3.0 p2)` that don’t exhibit this problem. I can recompile them and check.

> Switching to `--orphan-handling=discard` seems to get past this problem, but I'm not sure if the resulting binaries work

I believe the modules will work, but the we ought to insert a `*(.note.*)` entry in the `/DISCARD/` section of the IOP LD script, here:

https://github.com/frno7/iopmod/blob/13732bffa01dd29f3289a5a340a2ff08401557d3/script/iop.ld#L54-L61

> still working on getting my system to boot :D

Cool! How far have you come?

**Comment by frno7** [comment]
By the way, are you running Gentoo testing versions of both packages, by chance? That would explain why my versions are somewhat older, as I’m on the stable ones. Please try commit a5d2360a4f8a09f6fcf4dc58f6c87bcd1f926bc8 on the [notefix](https://github.com/frno7/iopmod/tree/notefix) branch.

**Comment by TobiX** [comment]
Yeah, that branch seems to work

**Comment by TobiX** [comment]
I tried to use `--stable` with crossdev, but it couldn't find a stable binutils (all current versions are masked?). Sorry, not much experience with Gentoo - see my [docker image](https://github.com/Dosage-Bot/mipsr5900el-toolchain)

**Comment by frno7** [comment]
The fix in commit a5d2360a4f8a09f6fcf4dc58f6c87bcd1f926bc8 is now on the master branch. Thanks for your report!
